mm2src/mm2_main/src/rpc.rs

```diff
@@ -314,62 +318,118 @@ pub extern "C" fn spawn_rpc(ctx_h: u32) {
     // then we might want to refactor into starting it ideomatically in order to benefit from a more graceful shutdown,
     // cf. https://github.com/hyperium/hyper/pull/1640.
 
+    let make_svc_fut = move |remote_addr: SocketAddr| async move {
+        Ok::<_, Infallible>(service_fn(move |req: Request<Body>| async move {
+            let res = rpc_service(req, ctx_h, remote_addr).await;
+            Ok::<_, Infallible>(res)
+        }))
+    };
+
+    //The `make_svc` macro creates a `make_service_fn` for a specified socket type.
+    // `$socket_type`: The socket type with a `remote_addr` method that returns a `SocketAddr`.
+    macro_rules! make_svc {
+        ($socket_type:ty) => {
+            make_service_fn(move |socket: &$socket_type| {
+                let remote_addr = socket.remote_addr();
+                make_svc_fut(remote_addr)
+            })
+        };
+    }
+
+    // The `get_shutdown_future` macro registers a graceful shutdown listener by calling the `register_listener`
+    // method of `GracefulShutdownRegistry`.
+    // If the `register_listener` method fails, it implies that the application is already in a shutdown state.
+    // In this case, the macro logs an error and immediately returns.
+    macro_rules! get_shutdown_future {
+        ($ctx:expr) => {
+            match $ctx.graceful_shutdown_registry.register_listener() {
+                Ok(shutdown_fut) => shutdown_fut,
+                Err(e) => {
+                    error!("MmCtx seems to be stopped already: {e}");
+                    return;
+                },
+            }
+        };
+    }
+
+    // Macro for spawning a server with error handling and logging
+    macro_rules! spawn_server {
+        ($server:expr, $ctx:expr, $ip:expr, $port:expr) => {
+            {
+                let server = $server.then(|r| {
+                    if let Err(err) = r {
+                        error!("{}", err);
+                    };
+                    futures::future::ready(())
+                });
+
+                // As it's said in the [issue](https://github.com/hyperium/tonic/issues/330):
+                //
+                // Aborting the server future will forcefully cancel all connections and not perform a proper drain/shutdown.
+                // While using the special shutdown methods on the server will allow hyper to gracefully drain all connections
+                // and gracefully close connections.
+                common::executor::spawn({
+                    log_tag!(
+                        $ctx,
+                        "😉";
+                        fmt = ">>>>>>>>>> DEX stats {}:{} DEX stats API enabled at unixtime.{} <<<<<<<<<",
+                        $ip,
+                        $port,
+                        gstuff::now_ms() / 1000
```

Comment:
why use `gstuff::now_ms()` here?

Reply:
Just copied it from the old implementation without considering it. I guess `now_sec()` can be used straight away instead.